diffsims/generators/simulation_generator.py

```diff
@@ -0,0 +1,422 @@
+                    f"{shape_factor_model} is not a recognized shape factor "
+                    f"model, choose from: {_shape_factor_model_mapping.keys()} "
+                    f"or provide your own function."
+                )
+        else:
+            self.shape_factor_model = shape_factor_model
+        self.minimum_intensity = minimum_intensity
+        self.shape_factor_kwargs = kwargs
+        if scattering_params in ["lobato", "xtables", None]:
+            self.scattering_params = scattering_params
+        else:
+            raise NotImplementedError(
+                "The scattering parameters `{}` is not implemented. "
+                "See documentation for available "
+                "implementations.".format(scattering_params)
+            )
+
+    @property
+    def wavelength(self):
+        return get_electron_wavelength(self.accelerating_voltage)
+
+    def calculate_diffraction2d(
+        self,
+        phase: Union[Phase, Sequence[Phase]],
+        rotation: Union[Rotation, Sequence[Rotation]] = Rotation.from_euler(
+            (0, 0, 0), degrees=True
+        ),
+        reciprocal_radius: float = 1.0,
+        with_direct_beam: bool = True,
+        max_excitation_error: float = 1e-2,
+        shape_factor_width: float = None,
+        debye_waller_factors: dict = None,
+    ):
+        """Calculates the diffraction pattern for one or more phases given a list
+        of rotations for each phase.
+
+        Parameters
+        ----------
+        phase:
+            The phase(s) for which to derive the diffraction pattern.
+        reciprocal_radius
+            The maximum radius of the sphere of reciprocal space to
+            sample, in reciprocal Angstroms.
+        rotation
+            The Rotation object(s) to apply to the structure and then
+            calculate the diffraction pattern.
+        with_direct_beam
+            If True, the direct beam is included in the simulated
+            diffraction pattern. If False, it is not.
+        max_excitation_error
+            The cut-off for geometric excitation error in the z-direction
+            in units of reciprocal Angstroms. Spots with a larger distance
+            from the Ewald sphere are removed from the pattern.
+            Related to the extinction distance and roughly equal to 1/thickness.
+        shape_factor_width
+            Determines the width of the reciprocal rel-rod, for fine-grained
+            control. If not set will be set equal to max_excitation_error.
+        debye_waller_factors
+            Maps element names to their temperature-dependent Debye-Waller factors.
+
+        Returns
+        -------
+        diffsims.sims.diffraction_simulation.DiffractionSimulation
+            The data associated with this structure and diffraction setup.
+        """
+        if isinstance(phase, Phase):
+            phase = [phase]
+        if isinstance(rotation, Rotation):
+            rotation = [rotation]
```

Comment:
If the lengths of `phase` or `rotation` are different, then the longer one gets silently ignored in the `zip` below. Maybe we should re-use the rotations if there are more phases. If there are more rotations than phases then I'm not so sure what is best.
```suggestion
        if isinstance(phase, Phase):
            phase = [phase]
        if isinstance(rotation, Rotation):
            rotation = [rotation] * len(phase)
```

Reply:
I just raised an error when this is the case. I think that is the best way to handle this without making it too complicated.